### DISCUSSION TITLE
Abbreviated journal title loses periods

### DISCUSSION THREAD

**Discussion by ryofurue** [retorquere/zotero-better-bibtex#1949] [Q&A] [answered]
In my Zotero database, some journal names are already abbreviated and others are not.  This is just for historical reasons. 

In the automatically generated bibtex file, 
- the original "Phil. Trans. Roy. Soc. A" is converted to "Phil Trans Roy Soc A";
- "Progress in Oceanography"  remains as is;
- "Journal of Geophysical Research, Oceans" is abbreviated to "J. Geophys. Res. Oceans"

The last example is as I expect and I like, but I need to fix the first example: I need the periods to stay.  Also, I want the second example to be abbreviated.

I have "Automatically abbreviate journal title if none is set explicitly" checked under "Better BibTex" > "Export" > "Miscellaneous".

Perhaps I have some wrong settings?

**Comment by retorquere** [comment]
Generated abbreviations come from the Zotero abbreviator; I don't generate them myself, I just pass them on. If you get the same abbreviations using "create bibliography from item", you'd have to talk to Zotero. If they differ, it's something I could look into.

> Perhaps I have some wrong settings?

I don't think so, as there's no setting I can think of that relates to this, but it's impossible to say without a support log. A support log would give me your current BBT settings, and a copy of the item under discussion. Without it I'd just be guessing.

**Reply by ryofurue** [reply]
The support log for one of the items: 4SLNIFQ8-refs-apse

Thank you for your help!  The "generate bibliography from item" feature gives the correct journal abbreviation with periods, whereas the periods are stripped out in the automatically-generated bibtex file:

- "Phil. Trans. Roy. Soc. A" --- On Zotero
- "Phil. Trans. Roy. Soc. A" ---  Right-click > "Generate bibliography from item . . . " > "Elsevier - Harvard (with titles)"
- "Phil Trans Roy Soc A" --- In the automatically-generated bibtex file.

**Reply by retorquere** [reply]
I have a build running that will try to detect your situation, but why do you have `Phil. Trans. Roy. Soc. A` in the publication field rather than in the journal abbreviation field? That's what threw BBT off.

**Reply by ryofurue** [reply]
> why do you have Phil. Trans. Roy. Soc. A in the publication field rather than in the journal abbreviation field?

I think those abbreviated names come from my old original bibtex file.  More than 95% of my entries are from my old bibtex file which I was using as the master database before switching to Zotero.  I imported it into Zotero as the first step of my migration to Zotero.

It's too late but in hindsight, the bibtex importer should have detected whether the original "journal" field was an abbreviation or not.

**Reply by retorquere** [reply]
That's "should" is not a trivial thing. How should it have done this?

My import does restore some abbreviations, but there isn't (to my knowledge) a reasonably complete list of journal abbreviations, and I would not know how to reliably detect that something is an abbreviation. If you can think of a decent heuristic, I can get you a script to run in Zotero's Javascript runner that will move the abbreviations into the journal abbreviation field in your database.

**Reply by ryofurue** [reply]
> How should it have done this?

I had something like this in mind . . . If the original "journal" field included a period, it would have been imported as Zotero's "Journal Abbr" and its "Publication" field would have remained empty.

**Reply by retorquere** [reply]
Can you post a sample of such an entry at #1951?

**Reply by retorquere** [reply]
To fix the existing items, you can select the problematic entries (perhaps by using search for a period in the journal name), and run this code in Zotero's Javascript runner. I've tested this locally, and the script tries to be conservative (no overwrite of existing fields, and verifies the abbrev was copied over), but you still have to make sure it does what you want by running small tests first. It will only inspect & attempt to modify selected entries.

```
for (const item of Zotero.getActiveZoteroPane().getSelectedItems()) {
  if (!item.isRegularItem()) continue
  const publicationTitle = item.getField('publicationTitle')
  const journalAbbreviation = item.getField('journalAbbreviation')
  if (!journalAbbreviation && publicationTitle && publicationTitle.includes('.')) {
    item.setField('journalAbbreviation', publicationTitle)
    if (item.getField('journalAbbreviation') === publicationTitle) { // make sure we copied it over
      item.setField('publicationTitle', '')
      await item.saveTx()
    }
  }
}
```

**Reply by retorquere** [reply]
> I had something like this in mind . . . If the original "journal" field included a period, it would have been imported as Zotero's "Journal Abbr" and its "Publication" field would have remained empty.

That conflicts with a few things that I have in my test suite as journal titles (provided by other BBT users):

* PublishersWeekly.com
* Multicultural perspectives (Mahwah, N.J.)
* Journal of African American studies (New Brunswick, N.J.)

and there's probably more like these

**Reply by retorquere** [reply]
I have some heuristics that will pick out the samples you provided which will be part of the next release, but these can never be complete; I must be very careful to not mess up imports for bib files that do have the abbreviations in the `shortjournal` field (which is always taken to be an abbreviation) before import, and that have journal titles with periods in the `journal` field.

**Reply by ryofurue** [reply]
Thank you for your careful discussion.   Now I start to wonder . . . 

What would happen if "PublishersWeekly.com" were imported into **both** `Journal Abbr` and the `Publication` fields?

I tested this idea.  I wrote "J. Phys. Oceanogr." in **both** `Journal Abbr` and `Publication` fields on Zotero.  In the auto-generated bibtex file, I found `journal = {J. Phys. Oceanogr.}`.  Yey!

So, what about importing the original `journal` field into the two Zotero fields if the original contains a period.

This solution isn't perfect [because we wouldn't be able to have "J. Afric. Amer. studies (New Brunswick, N.J.)" for "Journal of African American studies (New Brunswick, N.J.)"], but that would be a somewhat lesser harm because the vast majority of the journal don't include a period and because having an unabbreviated journal name is better than having an abbreviated name without periods, such as "Phil Trans R Soc".

Just my two cents.

**Reply by retorquere** [reply]
The new changes try to make sure that it is somewhat likely to be an abbreviation before moving it there; the rule in the new release will be "at least two periods, no periods that are followed by a non-space character, and no spaces that are not preceded by a period". But at the end of the day, I can't second-guess the user doing the import too much, given that the user has no way to steer this second-guessing; the text being in the `journal` is a pretty strong hint that it should be taken to be the journal name. "J. Afric. Amer. studies (New Brunswick, N.J.)" will be taken not to be an abbreviation, unless it is in `shortjournal`, in which case it will always taken to be an abbreviation.

**Comment by retorquere** [comment, marked as answer]
BBT was abbreviating journal abbreviations that were in the publication field rather than the journal abbreviation field. Moving them there will fix it.

**Reply by retorquere** [reply]
(this repeats the earlier resolution because I can only mark toplevel posts as answer)
